src/backend/distributed/planner/local_distributed_join_planner.c

```diff
@@ -0,0 +1,566 @@
+#include "utils/lsyscache.h"
+
+typedef struct RTEToSubqueryConverterReference
+{
+	RangeTblEntry *rangeTableEntry;
+	Index rteIndex;
+	List *restrictionList;
+	List *requiredAttributeNumbers;
+} RTEToSubqueryConverterReference;
+
+typedef struct RTEToSubqueryConverterContext
+{
+	List *distributedTableList; /* reference or distributed table */
+	List *localTableList; /* local or citus local table */
+	bool hasSubqueryRTE;
+}RTEToSubqueryConverterContext;
+
+static Oid GetResultRelationId(Query *query);
+static Oid GetRTEToSubqueryConverterReferenceRelId(
+	RTEToSubqueryConverterReference *rteToSubqueryConverterReference);
+static bool ShouldConvertLocalTableJoinsToSubqueries(List *rangeTableList, Oid
+													 resultRelationId);
+static bool HasUniqueFilter(RangeTblEntry *distRTE, List *distRTERestrictionList,
+							List *requiredAttrNumbersForDistRTE);
+static bool ShouldConvertDistributedTable(FromExpr *joinTree,
+										  RTEToSubqueryConverterReference *distRTEContext);
+static List * RequiredAttrNumbersForRelation(RangeTblEntry *relationRte,
+											 RecursivePlanningContext *planningContext);
+static RTEToSubqueryConverterContext * CreateRTEToSubqueryConverterContext(
+	RecursivePlanningContext *context,
+	List *
+	rangeTableList);
+static void GetAllUniqueIndexes(Form_pg_index indexForm, List **uniqueIndexes);
+static RTEToSubqueryConverterReference * GetNextRTEToConvertToSubquery(FromExpr *joinTree,
+																	   RTEToSubqueryConverterContext
+																	   *
+																	   rteToSubqueryConverterContext,
+																	   PlannerRestrictionContext
+																	   *
+																	   plannerRestrictionContext,
+																	   Oid
+																	   resultRelationId);
+static void GetRangeTableEntriesFromJoinTree(Node *joinNode, List *rangeTableList,
+											 List **joinRangeTableEntries);
+static void RemoveFromRTEToSubqueryConverterContext(
+	RTEToSubqueryConverterContext *rteToSubqueryConverterContext, Oid relationId);
+static bool FillLocalAndDistributedRTECandidates(
+	RTEToSubqueryConverterContext *rteToSubqueryConverterContext,
+	RTEToSubqueryConverterReference **
+	localRTECandidate,
+	RTEToSubqueryConverterReference **
+	distributedRTECandidate);
+
+/*
+ * ConvertUnplannableTableJoinsToSubqueries gets a query and the planner
+ * restrictions. As long as the query is not plannable by router planner,
+ * it converts either a local or distributed table to a subquery.
+ */
+void
+ConvertUnplannableTableJoinsToSubqueries(Query *query,
+										 RecursivePlanningContext *context)
+{
+	List *rangeTableList = NIL;
+	GetRangeTableEntriesFromJoinTree((Node *) query->jointree, query->rtable,
+									 &rangeTableList);
+	Oid resultRelationId = GetResultRelationId(query);
+	if (!ShouldConvertLocalTableJoinsToSubqueries(rangeTableList, resultRelationId))
+	{
+		return;
+	}
+
+	RTEToSubqueryConverterContext *rteToSubqueryConverterContext =
+		CreateRTEToSubqueryConverterContext(
```

Comment:
Is there a better name for this than "context"? We usually use context to pass down information when using recursion. In this case it seems we just want to reduce the number of arguments we're passing to some very specific functions. Can we give a descriptive name to the thing we're passing to this function? e.g. ConversionCandidates?

Reply:
ConversionCandidates sounds good to me, Onder was also a bit hesitant on this, as in `do we need this at all`. I am more inclined towards `ConversionCandidates` than not having this since the checks aren't that simple